### ISSUE TITLE
Backend Breaks With SSL

### ISSUE THREAD

**Issue by flip4bytes** [octobercms/october#856]
I've been looking for support and a way to fix this for a few days now with no avail.

The assets that are being called when you visit /backend are all called in via http://

This causes it to be broken as the browser won't load insecure assets over https://

The fix I did in Laravel does not work in OctoberCMS and making any forcing changes within the backend files just cause it to reset anyway when you run composer (of which happens every time you push your updated site to Heroku).

Howcome it even lists the complete directory by default on October anyway? Rather than listing "http://example.com/modules/foo/bar/.css&.js" why not have it by default just be "/modules/foo/bar/.css&.js" and then you wouldn't even have to deal with the native assets switching between http:// and https://.

**Comment by daftspunk** [comment]
This is generally inconsistent across the platform, I think because Laravel nor October have an opinion on this. Maybe we need to discuss the idea of a link policy? 

- relative: All links are created relative, hostname is omitted
- secure: All links are generated to the detected hostname and are HTTPS
- insecure: All links are generated to the detected hostname and are HTTP
- force[secure|insecure]: All links are generated to a specified hostname (see config `app.url`)

**Comment by flip4bytes** [comment]
Interesting. Yeah, I would be curious to hear what others have to say on the matter.

I have always been a large proponent in the past of having a relative link in all of my files. Not to mention, OctoberCMS already does that when embedding with /combine/ and various other assets on the front-end.

I have only since run into this problem in general since switching to a framework which has made dealing with moving a site over to SSL a real pain in the ass. I am now left with a problem for no reason simply due to links not being relative.

It also appears much cleaner on the back-end and you don't run into annoying problems where you need a separate `app.url` for localhost as for dev/production server, one forcing https and the other forcing http.

I guess the real question would be, when would you ever want to force say insecure assets over a secure (https) connection? Or when would you want to force secure assets over an insecure (http) connection? I can't think of a possibly time where someone would purposely want to do this which is why in my eyes, the 2nd & 3rd option doesn't make much sense to me. But maybe someone else can enlighten me?

As for the 4th option, pretty sure that's what Laravel does out of the box, but October seems to break (for the backend atleast). I have switched the `app.url` between http and https without any effect in the local asset management atleast. But even so, you're dealing with needing a separate local environment config app.php file and one for production, so that's creating an unnecessary need for someone to set up two environments which would not need to even exist if it just smartly recognized it relatively without the hostname even there.

TL;DR: I am completely sold on the relative path, option number 1 unless someone else could in some way educate me on why that's for some reason a terrible idea? (I really don't feel like someone is going to fight against relative because thinking logically, it just makes no sense to force one over the other instead of having it just automatically decide). The relative path does all the thinking for the user and it would reduce the barrier of entry for people who are new to PHP having to learn and fix why their site completely breaks upon switching over to an SSL connection.

Really appreciate you taking this point seriously @daftspunk - really enjoying OctoberCMS & recommending it to every developer I know :)

**Comment by JanvanCasteren** [comment]
One plus for this from me. I am making my first (personal) site with Octobercms and I really like it very much. 
In two months I will start working on a serious webapplication for a client. I am coming from Kohana and planned to make the switch to Laravel. But Octobercms seems so flexible and powerfull to me, that it makes a good candidate to make that application with. But only if it will be possible to use the backend on https.

**Comment by flip4bytes** [comment]
Yeah, I just logged out of the backend and can't even log back into my site unless I go to CloudFlare, disable SSL, visit my site in http:// and then login that way before finally turning back on https:// and accepting the browser to load unsafe assets.

True SSL is pretty much useless until the backend assets are called in relative which is becoming more and more important now that SSL even on generic sites is highly recommended by Google.

Looking to use OctoberCMS on my client's site that's coming up in a couple weeks so really hope it gets settled by then.

Cheers!

**Comment by StudioLE** [comment]
I'm being debilitated this error. October CMS / Laravel don't seem to realise that the site is running on SSL.  `URL::asset()` and `Backend::skinAsset()` return `http://` urls rather than `https://`

I'm using Ubuntu, Nginx & CloudFlare Flexible SSL. I assume it's the Flexible SSL causing the issue?

> Flexible SSL: There is an encrypted connection between your site visitors and CloudFlare, but not from CloudFlare to your server.
> [More information on CloudFlare](https://www.cloudflare.com/ssl)

**Comment by flip4bytes** [comment]
@StudioLE yeah exactly. I'm in the same position using CloudFlare. I'm sure that's it, you bring up a good point. It's definitely due to the SSL to CloudFlare but not to the server  because I tried like 20 different things but the site never actually realized it was on SSL and adjusted assets accordingly.

The easiest fix this would be to force everything from it's relative directory path as mentioned above because you wouldn't need to deal with http vs https URLs at all from local assets. But that's entirely up to the framework doing it this way. It'd be a huge pain to do it yourself as it'd be reverted next update or everything composer updates the backend. Kinda at a standstill. I guess all we can do is explain to the dev of how big of a problem this really is.

**Comment by MichaelBanks** [comment]
Same issue here. Cloudflare flexi SSL.

![](https://i.starbs.net/evn3)

**Comment by MichaelBanks** [comment]
I produced a terribly patchy workaround by tampering with Laravel's UrlGenerator. It was the quickest fix for me.

Rename the URL::to function to unsecureTo and force any uses of to to be secure. Such as URL::secureTo
Same goes for assets. I apologize for how cringeworthy this fix is, but it works.

https://i.starbs.net/A8dM

A production fix for OctoberCMS would be to create it's own functions for deciding if to use URL::to or URL::secureTo and URL::secureAsset over URL::asset. This would be achieved with a 'secure' (true/false) config option and/or checking if the URL issued in the config starts with `https`

**Comment by flip4bytes** [comment]
@MichaelBanks interesting. You mention the production fix is to check the config and see if it starts with https or not. Wouldn't a better production fix just be to output a relative url though?

e.g. instead of

`<link src="https://www.example.com/directory/to/backend/assets">`

have it just do

`<link src="/directory/to/backend/assets">`

Because with that fix, you don't have to deal with true/false or specify in the config if your site will be on SSL or not. I mean relative routes are the perfect fix for this in theory, is it not? I honestly don't see a downside..

It's the same reason that links like `<link src="//cdnjs.cloudflare.com/ajax/libs/jquery/2.1.3/jquery.min.js">` exist.. Just using `//` takes it upon the browser to choose whether it needs a secure or insecure asset. It functions just as a relative link works.

So from your suggestion for the production fix, I ask why even use unnecessary PHP statements and configs to define whether an asset is secure or not when it's not even needed because browsers have done that natively since the begging of time if you use the `relative` or `//` route.

**Comment by MichaelBanks** [comment]
@flip4bytes I do agree rel links would be a nice feature. But, I'd prefer if I could configure to use relative links/definite links and `//` links

**Comment by flip4bytes** [comment]
@MichaelBanks just curious if you can come up with even one scenario where it makes sense to configure it? Not trying to be argumentative, just trying to spark a discussion and save @daftspunk time having to create unnecessary options.

Can you come up with an example of where the developer would ever want to force secure server assets over an insecure connection or force insecure server assets over a secure connection? I've been thinking about it for months now and can't come up with even one example where that would EVER make sense. Therefore it would be a waste of time implementing anything but the relative path option. Just saying what you prefer doesn't give it the proper backing and I feel like @daftspunk is waiting for an actual discussion of the best route before fixing this bug.

Interested in hearing your ideas.

**Comment by MichaelBanks** [comment]
@flip4bytes the only purpose I can think from the top of my head is shifting assets off-site.

**Comment by flip4bytes** [comment]
@MichaelBanks but this suggestion is just for how on-site assets are handled anyway. Calling in off-site assets would still work the same as it always has, just linking to the whole URL and choosing then if it's http or https. This is only about how on-site assets are handled, aka relative still makes the most sense.

**Comment by MichaelBanks** [comment]
@flip4bytes I agree, I can't imagine what I was thinking the first time around.
Nice to see @daftspunk making progress on this issue :+1:

**Comment by flip4bytes** [comment]
@MichaelBanks appreciate the discussion though! And @daftspunk thank you so much for moving forward on this! Can't want for the update, excited to start using the /backend again without having to first disable CloudFlare SSL. (:

**Comment by MichaelBanks** [comment]
https://github.com/octobercms/october/commit/47eff923e5b8a7a90bb931fac40628938ff16678
Policy reverted due to breaks.

@daftspunk can you elaborate?

**Comment by flip4bytes** [comment]
@daftspunk @MichaelBanks So I have updated to build 230 and the problem still exists. `detect` mode does not work properly if you are just using CloudFlare's Flexible SSL as per @StudioLE mentioned:

>Flexible SSL: There is an encrypted connection between your site visitors and CloudFlare, but not from CloudFlare to your server.

I guess this is a perfect example of why having a relative link policy is preferred. The server is seeing it as not SSL, but the site visitor is seeing that it as SSL. Because of that, October with the `detect` policy in place detects the site as non-SSL when in reality it's being delivered as SSL. Using a relative link policy is the only way to fix that. Unless of course you forced a secure connection, but then in your local environment, you also have to deal with a secure connection being forced to deliver the assets which is a complete pain in the butt.

**Comment by serversides** [comment]
I am having the same issue.  Would be nice to be able to fix without any extreme tinkering.

**Comment by dylanjacob** [comment]
+1 from me too.  SSL on the backend is a biggie for me to feel confident in using October CMS as my goto for clients.

**Comment by itsphi1989** [comment]
Adding this code to my .htaccess forces redirects everything to https including backend.

<code>
RewriteCond %{HTTPS} !=on
</code>
<code>
RewriteRule ^/?(.*) https://%{SERVER_NAME}/$1 [R,L]
</code>

 Let me know if this helped anyone :/

**Comment by toome123** [comment]
It's about one year ago, have any possible fix ?

**Comment by StudioLE** [comment]
@toome123 If I recall correctly, in `/config/cms.php` you can now force HTTPS links by setting:

```php
'linkPolicy' => 'secure'
```

**Comment by abass** [comment]
@toome123 I just avoided all of that mess by getting an SSL certificate. If you are using Laravel Forge, they just integrated "Auto-renewing Let's Encrypt SSL certificates" which are 100% FREE SSL Certs and it makes it incredibly easy. Just click "SSL Cert --> LetsEncrypt --> Activate" then you can just force full SSL from CloudFlare and you're good to go. Takes not even 40 seconds to get up and working on each site.

**Comment by ayumihamsaki** [comment]
@StudioLE just tried: 'linkPolicy' => 'secure' in /config/cms.php does not for me. HTTP and HTTPs both working and no redirects.

**Comment by daftspunk** [comment]
This only controls how links are generated, it does not enforce any redirects as this would be too magical. This is up to your web server configuration to handle.

**Comment by ayumihamsaki** [comment]
@daftspunk thanks for quick reply. This is probably me being stupid but its not working with this in my htaccess:

```
RewriteEngine On
RewriteCond %{HTTPS} off
RewriteCond %{HTTP_HOST} ^domain.com [NC,OR]
RewriteCond %{HTTP_HOST} ^www.domain.com [NC]
RewriteRule ^(.*)$ https://www.domain.com/$1 [L,R=301]
```

Can we not do redirects in htaccess with octobercms (sorry I only started using your wonderful cms 24 hours ago)

**Comment by Eoler** [comment]
This works well with Apache:

**OctoberCMS config:**
```
    'cms.linkPolicy' => 'force',
    'app.url' => 'https://...',
```
**.htaccess**
```
    ## Force secure connection
    RewriteCond %{HTTP:X-Forwarded-Proto} =http [OR]
    RewriteCond %{HTTP:X-Forwarded-Proto} =""
    RewriteCond %{HTTPS} !=on
    RewriteRule ^ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]
```

**Comment by srsgores** [comment]
For me, setting `cms.linkPolicy` to `'secure'` did the trick.  Also, it fixed the responsiveImages plugin from having the same problem

**Comment by Samuell1** [comment]
Best simple fix is to force SSL on server or dns side and dont forget to disable in _config/cms.php_ **backendForceSecure** to `false`

**Comment by ayumihamsaki** [comment]
Not sure why Github has linked me to this issue?

I will just add if you use `Cloudflare` and` October CMS`, Eoler's solution to add the code to` .HTaccess` won't always work and you should create a custom `Page Rules`, for more info see here: https://support.cloudflare.com/hc/en-us/articles/204144518
